File: Backend/src/models/telephone_numbers.py
# Backend/src/models/telephone_numbers.py
# Definition of the TelephoneNumber model

# Author: Valentin Haas, 2024

# System imports
from datetime import datetime
import enum

# Library imports
from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_pascal
from sqlalchemy import Column, DateTime, Enum, Integer, String

# Project imports
from ._base import BaseTable


class TelephoneNumberType(enum.Enum):
    """The type of the telephone number."""

    OTHER = 0
    """The telephone number is of another type."""

    HOME = 1
    """The telephone number is a home number."""

    MOBILE = 2
    """The telephone number is a mobile number."""

    WORK = 3
    """The telephone number is a work number."""

    FAX = 4
    """The telephone number is a fax number."""


class TelephoneNumbersTable(BaseTable):
    """Definition of the TelephoneNumber model for the database."""

    __tablename__ = "TelephoneNumbers"
    Id = Column(Integer, primary_key=True)
    """The unique identifier of the telephone number."""

    Number = Column(String, nullable=False)
    """The number of the telephone number."""

    Label = Column(String, nullable=False)
    """The label of the telephone number."""

    Type = Column(Enum(TelephoneNumberType), nullable=False)
    """The type of the telephone number."""

    CreationDate = Column(DateTime, nullable=False, default=datetime.now())
    """The creation date of the telephone number."""

    def to_pydantic(self):
        return super().to_pydantic(self, TelephoneNumber)


class TelephoneNumber(BaseModel):
    """Definition of the TelephoneNumber model for the API."""

    model_config = ConfigDict(populate_by_name=True, alias_generator=to_pascal)

    id: int
    """The unique identifier of the telephone number."""

    number: str
    """The number of the telephone number."""

    label: str
    """The label of the telephone number."""

    type: TelephoneNumberType
    """The type of the telephone number."""

    creation_date: datetime
    """The creation date of the telephone number."""

File: Backend/src/models/telephonenumbers.go
// Backend/src/models/telephonenumbers.go
// Definition of the TelephoneNumber model

// Author: Valentin Haas, 2025
package models

import (
	"gorm.io/gorm"
)

// TelephoneNumberType defines the type enumeration for telephone numbers.
type TelephoneNumberType int

const (
	OTHER_TEL_NUMBER_TYPE TelephoneNumberType = iota // The telephone number is of another type.
	HOME                                             // The telephone number is a home number.
	MOBILE                                           // The telephone number is a mobile number.
	WORK                                             // The telephone number is a work number.
	FAX                                              // The telephone number is a fax number.
)

// TelephoneNumbers defines the TelephoneNumber model for the database.
type TelephoneNumber struct {
	gorm.Model                     // Provides ID, CreatedAt, UpdatedAt, DeletedAt fields
	Number     string              `gorm:"not null"` // The number of the telephone number.
	Label      string              `gorm:"not null"` // The label of the telephone number.
	Type       TelephoneNumberType `gorm:"not null"` // The type of the telephone number.
}
